File: src/scripts/lint_command_verbs.py
#!/usr/bin/env python3
"""Controlled-verb linter for visible commands (6.0.0-C Phase 2 Step 4; ADR-041).

A VISIBLE command (tier 0/1) must have a leading token drawn from the approved
verb allowlist in ``src/config/discovery/command-verbs.yml``. ``create-*`` is a
banned leading token for new visible commands (``create-pr`` is grandfathered).

FORWARD-ONLY: only command files **added** since ``--baseline`` (default
``main``), plus files **promoted** to a visible tier since baseline, are
checked. The existing surface is grandfathered — no rename wave. Mirrors the
baseline-diff machinery of ``lint_no_new_atomic_commands.py``.

Single modular lint: both the approved-verb rule and the banned-prefix rule run
in one pass. Every rule blocks CI (no advisory-only tier — see ADR-041 § 6).

Exit codes: 0 = clean, 1 = violations found, 3 = internal error.

Usage:
    python3 scripts/lint_command_verbs.py
    python3 scripts/lint_command_verbs.py --baseline origin/main
    python3 scripts/lint_command_verbs.py --all     # check every visible command
"""

from __future__ import annotations

import argparse
import re
import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent.parent
VERBS_YML = ROOT / "src" / "config" / "discovery" / "command-verbs.yml"
_CMD_PATH_RE = re.compile(r"\.agent-src\.uncondensed/commands/.+\.md$")
NAME_RE = re.compile(r"^name:\s*(.*)$", re.MULTILINE)
TIER_RE = re.compile(r"^tier:\s*(\d+)", re.MULTILINE)
VISIBLE_TIERS = {0, 1}


@dataclass
class Violation:
    file: str
    rule: str
    reason: str


def load_config() -> tuple[set[str], set[str], set[str]]:
    import yaml
    data = yaml.safe_load(VERBS_YML.read_text(encoding="utf-8")) or {}
    approved = set(data.get("approved_verbs") or [])
    banned = set(data.get("banned_prefixes") or [])
    grandfathered = set(data.get("grandfathered") or [])
    if not approved:
        print(f"❌  No approved_verbs in {VERBS_YML}", file=sys.stderr)
        sys.exit(3)
    return approved, banned, grandfathered


def _git(args: list[str], *, tolerant: bool = False) -> str:
    try:
        r = subprocess.run(["git", *args], capture_output=True, text=True,
                           cwd=ROOT, timeout=15)
    except (FileNotFoundError, subprocess.TimeoutExpired) as exc:
        print(f"❌  git {' '.join(args)} failed: {exc}", file=sys.stderr)
        sys.exit(3)
    if r.returncode != 0:
        if tolerant:
            return ""
        print(f"❌  git {' '.join(args)} exit {r.returncode}: {r.stderr}",
              file=sys.stderr)
        sys.exit(3)
    return r.stdout


def _parse(text: str) -> tuple[str | None, int]:
    """(name, tier) from frontmatter text; tier defaults to 2 (internal)."""
    nm = NAME_RE.search(text)
    name = nm.group(1).strip().strip('"').strip("'") if nm else None
    tm = TIER_RE.search(text)
    tier = int(tm.group(1)) if tm else 2
    return name, tier


def changed_command_files(baseline: str) -> dict[str, str]:
    """relpath → change kind ('A' added | 'M' modified) for command files."""
    out: dict[str, str] = {}
    for kind in ("A", "M"):
        for p in _git(["diff", "--name-only", f"--diff-filter={kind}",
                       f"{baseline}...HEAD"]).splitlines():
            if p.strip() and _CMD_PATH_RE.search(p):
                out[p.strip()] = kind
    for line in _git(["status", "--porcelain", "-uall"]).splitlines():
        st, path = line[:2].strip(), line[3:].strip().split(" -> ")[-1]
        if _CMD_PATH_RE.search(path):
            out[path] = "A" if st in ("A", "??", "AM") else out.get(path, "M")
    return out


def all_visible_command_files() -> dict[str, str]:
    out: dict[str, str] = {}
    pkgs = ROOT / "packages"
    roots = [d for d in pkgs.glob("*/.agent-src.uncondensed/commands")
             if d.is_dir()] if pkgs.is_dir() else []
    for root in roots:
        for md in root.rglob("*.md"):
            if md.name == "AGENTS.md" or "_archive" in md.parts:
                continue
            out[str(md.relative_to(ROOT))] = "A"
    return out


def leading_token(name: str) -> str:
    """The verb token: cluster sub-name's head, or the bare name's head."""
    sub = name.split(":")[-1]
    return sub.split("-")[0]


def check(relpath: str, kind: str, baseline: str, approved: set[str],
          banned: set[str], grandfathered: set[str]) -> list[Violation]:
    abs_path = ROOT / relpath
    if not abs_path.exists():
        return []  # deleted
    name, tier = _parse(abs_path.read_text(encoding="utf-8"))
    if name is None or tier not in VISIBLE_TIERS:
        return []  # internal / unnamed — not gated
    if kind == "M":
        # Only a PROMOTION into visibility counts as a new visible surface.
        prev = _git(["show", f"{baseline}:{relpath}"], tolerant=True)
        if prev:
            _, prev_tier = _parse(prev)
            if prev_tier in VISIBLE_TIERS:
                return []  # already visible before — grandfathered

    if name in grandfathered:
        return []  # documented single-command exception — exempt from both rules
    bare = name.split(":")[-1]
    vio: list[Violation] = []
    # Rule 1 — banned prefix (create-*).
    for bp in banned:
        if bare == bp or bare.startswith(bp + "-"):
            vio.append(Violation(relpath, "banned-prefix",
                                 f"`{name}` uses the banned leading token "
                                 f"`{bp}` (no {bp}-* commands — ADR-041 § 2). "
                                 f"Grandfather it in command-verbs.yml only with "
                                 f"a documented exception."))
            return vio  # banned message is the actionable one; don't pile on
    # Rule 2 — approved verb.
    tok = leading_token(name)
    if tok not in approved:
        vio.append(Violation(relpath, "approved-verb",
                             f"leading token `{tok}` of `{name}` is not an "
                             f"approved verb. Rename to an existing verb, or add "
                             f"`{tok}` to src/config/discovery/command-verbs.yml in "
                             f"its own PR with an ADR (ADR-041 § 5)."))
    return vio


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--baseline", default="main",
                    help="git ref to diff against (default: main)")
    ap.add_argument("--all", action="store_true",
                    help="check every visible command, not just changed ones")
    ap.add_argument("--quiet", action="store_true")
    args = ap.parse_args()

    approved, banned, grandfathered = load_config()
    targets = (all_visible_command_files() if args.all
               else changed_command_files(args.baseline))
    if not targets:
        if not args.quiet:
            print(f"✅  No new/changed commands under commands/ "
                  f"(baseline: {args.baseline}).")
        return 0

    violations: list[Violation] = []
    for relpath, kind in sorted(targets.items()):
        violations += check(relpath, kind, args.baseline, approved,
                            banned, grandfathered)

    if violations:
        print(f"❌  {len(violations)} controlled-verb violation(s):")
        for v in violations:
            print(f"  • [{v.rule}] {v.file} — {v.reason}")
        print("\nSee docs/decisions/ADR-041-controlled-command-verbs.md.")
        return 1
    if not args.quiet:
        print(f"✅  {len(targets)} new/changed command(s) all use an approved "
              f"verb (baseline: {args.baseline}).")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/scripts/lint_command_verbs.ts
#!/usr/bin/env tsx
/**
 * Controlled-verb linter for visible commands.
 *
 * TypeScript twin of `src/scripts/lint_command_verbs.py` (ADR-088,
 * Phase 4 / Wave 4b). The CLI contract is mirrored EXACTLY — `--baseline`
 * / `--all` / `--quiet` flags, exit codes (0 clean, 1 violations, 3
 * internal error), stdout/stderr split, byte-identical finding messages,
 * same scan scope (forward-only git diff + working-tree status, or every
 * visible command under `--all`), and the same sorted iteration order.
 * snake_case kept. No behaviour changes — latent bugs replicated.
 *
 * A VISIBLE command (tier 0/1) must have a leading token drawn from the
 * approved verb allowlist in `src/config/discovery/command-verbs.yml`.
 * `create-*` is a banned leading token for new visible commands
 * (`create-pr` is grandfathered).
 *
 * Exit codes: 0 = clean, 1 = violations found, 3 = internal error.
 */

import { spawnSync } from 'node:child_process';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { parse as parseYaml } from 'yaml';

const _HERE = fileURLToPath(import.meta.url);
// src/scripts/lint_command_verbs.ts → three dirs up is the repo root.
// Mirrors Python `Path(__file__).resolve().parent.parent.parent`.
const ROOT = path.resolve(path.dirname(_HERE), '..', '..');
const VERBS_YML = path.join(ROOT, 'src', 'config', 'discovery', 'command-verbs.yml');
const _CMD_PATH_RE = /\.agent-src\.uncondensed\/commands\/.+\.md$/;
const NAME_RE = /^name:\s*(.*)$/m;
const TIER_RE = /^tier:\s*(\d+)/m;
const SUB_RE = /^sub:\s*(.*)$/m;
const VISIBLE_TIERS: ReadonlySet<number> = new Set([0, 1]);

interface Violation {
    file: string;
    rule: string;
    reason: string;
}

function _exists(p: string): boolean {
    try {
        fs.statSync(p);
        return true;
    } catch {
        return false;
    }
}

/** POSIX relative path of `target` under `root` (str(Path.relative_to)). */
function _relTo(target: string, root: string): string {
    return path.relative(root, target).split(path.sep).join('/');
}

/** Strip one leading and one trailing matching quote pair, mirroring Python
 * `.strip('"').strip("'")` chains applied to a value. */
function _stripQuotes(value: string): string {
    let v = value;
    // Python `.strip('"')` removes ALL leading/trailing double quotes, then
    // `.strip("'")` removes ALL leading/trailing single quotes.
    v = v.replace(/^"+/, '').replace(/"+$/, '');
    v = v.replace(/^'+/, '').replace(/'+$/, '');
    return v;
}

function load_config(): { approved: Set<string>; banned: Set<string>; grandfathered: Set<string> } {
    let raw: string;
    try {
        raw = fs.readFileSync(VERBS_YML, 'utf-8');
    } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        process.stderr.write(`❌  No approved_verbs in ${VERBS_YML}\n`);
        // Match Python: a load failure surfaces as missing approved_verbs.
        void msg;
        process.exit(3);
    }
    const data = (parseYaml(raw, { version: '1.1' }) as Record<string, unknown> | null) ?? {};
    const approved = new Set<string>((data['approved_verbs'] as string[] | null) ?? []);
    const banned = new Set<string>((data['banned_prefixes'] as string[] | null) ?? []);
    const grandfathered = new Set<string>((data['grandfathered'] as string[] | null) ?? []);
    if (approved.size === 0) {
        process.stderr.write(`❌  No approved_verbs in ${VERBS_YML}\n`);
        process.exit(3);
    }
    return { approved, banned, grandfathered };
}

function _git(args: string[], tolerant = false): string {
    const r = spawnSync('git', args, {
        cwd: ROOT,
        encoding: 'utf8',
        timeout: 15000,
    });
    if (r.error) {
        // FileNotFoundError / timeout equivalent.
        process.stderr.write(`❌  git ${args.join(' ')} failed: ${r.error.message}\n`);
        process.exit(3);
    }
    if (r.status !== 0) {
        if (tolerant) {
            return '';
        }
        process.stderr.write(`❌  git ${args.join(' ')} exit ${r.status}: ${r.stderr}\n`);
        process.exit(3);
    }
    return r.stdout;
}

/** (name, tier, sub) from frontmatter text; tier defaults to 2 (internal). */
function _parse(text: string): { name: string | null; tier: number; sub: string | null } {
    const nm = NAME_RE.exec(text);
    const name = nm ? _stripQuotes(nm[1]!.trim()) : null;
    const tm = TIER_RE.exec(text);
    const tier = tm ? parseInt(tm[1]!, 10) : 2;
    const sm = SUB_RE.exec(text);
    const sub = sm ? _stripQuotes(sm[1]!.trim()) : null;
    return { name, tier, sub };
}

function changed_command_files(baseline: string): Record<string, string> {
    const out: Record<string, string> = {};
    for (const kind of ['A', 'M'] as const) {
        const diff = _git([
            'diff',
            '--name-only',
            `--diff-filter=${kind}`,
            `${baseline}...HEAD`,
        ]);
        for (const line of diff.split('\n')) {
            const p = line.trim();
            if (p && _CMD_PATH_RE.test(p)) {
                out[p] = kind;
            }
        }
    }
    const status = _git(['status', '--porcelain', '-uall']);
    for (const line of status.split('\n')) {
        if (line === '') {
            continue;
        }
        const st = line.slice(0, 2).trim();
        const filePart = line.slice(3).trim().split(' -> ').pop()!;
        if (_CMD_PATH_RE.test(filePart)) {
            out[filePart] =
                st === 'A' || st === '??' || st === 'AM' ? 'A' : (out[filePart] ?? 'M');
        }
    }
    return out;
}

/** Recursively list `*.md` files under `dir`, mirroring Path.rglob('*.md'). */
function _rglobMd(dir: string): string[] {
    const out: string[] = [];
    const walk = (current: string): void => {
        let entries: fs.Dirent[];
        try {
            entries = fs.readdirSync(current, { withFileTypes: true });
        } catch {
            return;
        }
        for (const entry of entries) {
            const full = path.join(current, entry.name);
            if (entry.isDirectory()) {
                walk(full);
            } else if (entry.isFile() && entry.name.endsWith('.md')) {
                out.push(full);
            }
        }
    };
    walk(dir);
    return out;
}

function all_visible_command_files(): Record<string, string> {
    const out: Record<string, string> = {};
    const pkgs = path.join(ROOT, 'packages');
    let roots: string[] = [];
    if (_exists(pkgs) && fs.statSync(pkgs).isDirectory()) {
        // packages/*/.agent-src.uncondensed/commands directories.
        let pkgDirs: fs.Dirent[];
        try {
            pkgDirs = fs.readdirSync(pkgs, { withFileTypes: true });
        } catch {
            pkgDirs = [];
        }
        for (const d of pkgDirs) {
            const cand = path.join(pkgs, d.name, '.agent-src.uncondensed', 'commands');
            if (_exists(cand) && fs.statSync(cand).isDirectory()) {
                roots.push(cand);
            }
        }
        roots = roots.sort();
    }
    for (const root of roots) {
        for (const md of _rglobMd(root)) {
            const parts = _relTo(md, ROOT).split('/');
            if (path.basename(md) === 'AGENTS.md' || parts.includes('_archive')) {
                continue;
            }
            out[_relTo(md, ROOT)] = 'A';
        }
    }
    return out;
}

function leading_token(name: string, sub: string | null = null): string {
    const base = sub ? sub : name.split(':').pop()!;
    return base.split('-')[0]!;
}

function check(
    relpath: string,
    kind: string,
    baseline: string,
    approved: Set<string>,
    banned: Set<string>,
    grandfathered: Set<string>,
): Violation[] {
    const abs_path = path.join(ROOT, relpath);
    if (!_exists(abs_path)) {
        return []; // deleted
    }
    const { name, tier, sub } = _parse(fs.readFileSync(abs_path, 'utf-8'));
    if (name === null || !VISIBLE_TIERS.has(tier)) {
        return []; // internal / unnamed — not gated
    }
    if (kind === 'M') {
        // Only a PROMOTION into visibility counts as a new visible surface.
        const prev = _git(['show', `${baseline}:${relpath}`], true);
        if (prev) {
            const { tier: prev_tier } = _parse(prev);
            if (VISIBLE_TIERS.has(prev_tier)) {
                return []; // already visible before — grandfathered
            }
        }
    }

    if (grandfathered.has(name)) {
        return []; // documented single-command exception — exempt from both rules
    }
    const bare = sub ? sub : name.split(':').pop()!;
    const vio: Violation[] = [];
    // Rule 1 — banned prefix (create-*).
    for (const bp of banned) {
        if (bare === bp || bare.startsWith(bp + '-')) {
            vio.push({
                file: relpath,
                rule: 'banned-prefix',
                reason:
                    `\`${name}\` uses the banned leading token ` +
                    `\`${bp}\` (no ${bp}-* commands — ADR-041 § 2). ` +
                    `Grandfather it in command-verbs.yml only with ` +
                    `a documented exception.`,
            });
            return vio; // banned message is the actionable one; don't pile on
        }
    }
    // Rule 2 — approved verb.
    const tok = leading_token(name, sub);
    if (!approved.has(tok)) {
        vio.push({
            file: relpath,
            rule: 'approved-verb',
            reason:
                `leading token \`${tok}\` of \`${name}\` is not an ` +
                `approved verb. Rename to an existing verb, or add ` +
                `\`${tok}\` to src/config/discovery/command-verbs.yml in ` +
                `its own PR with an ADR (ADR-041 § 5).`,
        });
    }
    return vio;
}

interface ParsedArgs {
    baseline: string;
    all: boolean;
    quiet: boolean;
}

function parse_args(argv: readonly string[]): ParsedArgs {
    let baseline = 'main';
    let all = false;
    let quiet = false;
    for (let i = 0; i < argv.length; i++) {
        const arg = argv[i]!;
        if (arg === '--baseline') {
            const v = argv[++i];
            if (v === undefined) {
                _argparse_error('argument --baseline: expected one argument');
            }
            baseline = v;
        } else if (arg.startsWith('--baseline=')) {
            baseline = arg.slice('--baseline='.length);
        } else if (arg === '--all') {
            all = true;
        } else if (arg === '--quiet') {
            quiet = true;
        } else if (arg === '-h' || arg === '--help') {
            process.stdout.write(
                'usage: lint_command_verbs.py [-h] [--baseline BASELINE] [--all] [--quiet]\n',
            );
            process.exit(0);
        } else {
            _argparse_error(`unrecognized arguments: ${arg}`);
        }
    }
    return { baseline, all, quiet };
}

function _argparse_error(message: string): never {
    // Mirror Python argparse: usage line + program error, both on stderr,
    // exit code 2. The prog name is the basename of the Python original.
    process.stderr.write(
        'usage: lint_command_verbs.py [-h] [--baseline BASELINE] [--all] [--quiet]\n',
    );
    process.stderr.write(`lint_command_verbs.py: error: ${message}\n`);
    process.exit(2);
}

function main(): number {
    const args = parse_args(process.argv.slice(2));

    const { approved, banned, grandfathered } = load_config();
    const targets = args.all ? all_visible_command_files() : changed_command_files(args.baseline);
    if (Object.keys(targets).length === 0) {
        if (!args.quiet) {
            process.stdout.write(
                `✅  No new/changed commands under commands/ (baseline: ${args.baseline}).\n`,
            );
        }
        return 0;
    }

    const violations: Violation[] = [];
    for (const relpath of Object.keys(targets).sort()) {
        const kind = targets[relpath]!;
        violations.push(...check(relpath, kind, args.baseline, approved, banned, grandfathered));
    }

    if (violations.length) {
        process.stdout.write(`❌  ${violations.length} controlled-verb violation(s):\n`);
        for (const v of violations) {
            process.stdout.write(`  • [${v.rule}] ${v.file} — ${v.reason}\n`);
        }
        process.stdout.write('\nSee docs/decisions/ADR-041-controlled-command-verbs.md.\n');
        return 1;
    }
    if (!args.quiet) {
        process.stdout.write(
            `✅  ${Object.keys(targets).length} new/changed command(s) all use an approved ` +
                `verb (baseline: ${args.baseline}).\n`,
        );
    }
    return 0;
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exit(main());
}

export {
    type Violation,
    ROOT,
    VERBS_YML,
    _CMD_PATH_RE,
    VISIBLE_TIERS,
    load_config,
    changed_command_files,
    all_visible_command_files,
    leading_token,
    check,
    main,
};
